### ISSUE TITLE
Pull Requests support in the same repository (between the branches)

### ISSUE THREAD

**Issue by bastiao** [gogits/gogs#1597]

**Comment by bhchew** [comment]
+1

**Comment by ewoutp** [comment]
+1

**Comment by moritzuehling** [comment]
+1

**Comment by main--** [comment]
:+1:

**Comment by marcinwyszynski** [comment]
+1, it's actually a blocker for me

**Comment by Tim91084** [comment]
+1

**Comment by carlosfvp** [comment]
+1

**Comment by aganoza** [comment]
@Unknwon please could you point me to the relevant files to help make this feature happen?, regardless thanks for your time making this project.

**Comment by Unknwon** [comment]
@aganoza it's as simple as edit a file to make this work.

**Comment by JohnMaguire** [comment]
:+1: 

And an aside: @Unknwon I am very thankful for Gogs. I've been running it a while now and it's been solid. I am glad that there is a real competitor to GitLab, and I'm glad it's in the Go ecosystem, You've been a champ pushing out this much code for the past year and a half. Seriously, thank you.

**Comment by Unknwon** [comment]
@JohnMaguire :trollface:

**Comment by chrispappas** [comment]
+1 for this - adding the ability to do "Merge Requests" ala Gitlab would be a great addition

**Comment by martindafonte** [comment]
It would be nice indeed, it would make it a lot easier to our team to implement code revisión with this feature

**Comment by Fedott** [comment]
+1

**Comment by drsect0r** [comment]
+1

**Comment by dumpstate** [comment]
+1

**Comment by daniviga** [comment]
+1

**Comment by dabaer** [comment]
+1 Yas, needed very much.

**Comment by mkaito** [comment]
:+1:

**Comment by osklyar** [comment]
:+1:

**Comment by mkaito** [comment]
This is the last thing I'm actually waiting on. Then I can begin pestering @Unknwon about trello integration :stuck_out_tongue:

**Comment by bhchew** [comment]
Think in conjunction, will also need to implement more fine grain access control within the repo. If not anyone can merge the changes, bypassing the PR... Great work @Unknwon !

**Comment by Unknwon** [comment]
:trollface:

**Comment by skorokithakis** [comment]
+1

**Comment by nestabur** [comment]
+1

**Comment by windsource** [comment]
+1

**Comment by BurakDev** [comment]
+1, I need it too :sunglasses:

**Comment by gnanakeethan** [comment]
+1

**Comment by doliver3** [comment]
Please add this feature for pull requests between branches in the same repo. This is a critical essential feature.

**Comment by mgracer48** [comment]
+1000000

**Comment by pohy** [comment]
+1, this is much needed

**Comment by kisst** [comment]
+1

**Comment by hermanbergwerf** [comment]
+1

**Comment by codingfishman** [comment]
+1

**Comment by randy-girard** [comment]
+1

**Comment by jlbribeiro** [comment]
*+1*.
(And thank you, @Unknwon, for the amazing work, Gogs is amazing! The memory footprint <3

**Comment by kotchuprik** [comment]
I need it! :fire_engine:

**Comment by ivanlemeshev** [comment]
+1

**Comment by markramakers** [comment]
+1

**Comment by pentix** [comment]
:+1: +1

**Comment by migibert** [comment]
+1

**Comment by beowulf13th** [comment]
+1 :)

**Comment by lucacervasio** [comment]
+1

**Comment by squishykid** [comment]
+1

**Comment by qunabu** [comment]
+1

**Comment by lostsnow** [comment]
+1

**Comment by lovrenca** [comment]
+1

**Comment by dmavrin** [comment]
+1

**Comment by galaktor** [comment]
This feature is very important if I want to get more people to use our gogs server.

**Comment by Techwolf12** [comment]
+1

**Comment by misugijunz** [comment]
Hi. As I post in your blog article's comment section. The scenario will be like this https://confluence.atlassian.com/bitbucketserver/using-pull-requests-in-bitbucket-server-776639997.html

a. There will be a master branch that will updated periodically.
b. There will be several dev branches. Let say we have R1 for release one dev branch.
c. Developer should create a new branch when working on a feature/bugfix against dev branch. For example, a new feature for R1 will be developed under 'feature/name_r1_featurename' (change feature to bug fix if it is bug fixing). He/she is committing and pushing the works to the organization repository.
d. Dev login the stash and create new pull request from his branch to existing dev branch and assign few peer devs to review it (ex: feature/name_r1_featurename to R1)
e. Assigned peer devs can review the changes made and acknowledge the pull request if there is no issue.
f. Team lead can merge it (since usual dev don't have capability to merge it).

**Comment by tobigue** [comment]
+1

**Comment by guillemcanal** [comment]
+1

**Comment by arnaudruffin** [comment]
+1

**Comment by antechrestos** [comment]
+1

**Comment by sksullivan** [comment]
+1

**Comment by mccordnate** [comment]
+1

**Comment by felipesere** [comment]
+1

**Comment by thomas15v** [comment]
:+1:

**Comment by rsids** [comment]
:+1:

**Comment by earlye** [comment]
+1

**Comment by getson** [comment]
+1

**Comment by stubobis1** [comment]
+1

**Comment by stevoh6** [comment]
@misugijunz really good explanation. This function is important to many developers :-)

+1

**Comment by vladimir-polyakov** [comment]
+1

**Comment by okv** [comment]
+1

**Comment by hydandata** [comment]
+1 This is an amazing feature, relying it a lot on GitHub.

**Comment by stubobis1** [comment]
Is there any update on this? Is it possible to do?

**Comment by Unknwon** [comment]
@stubobis1 not spossible yet.

**Comment by stubobis1** [comment]
@Unknwon what sort of changes need to take place?

**Comment by Unknwon** [comment]
@stubobis1 thanks for asking but it's a false question, because you're going to have no ideas about what I'm talking in technical details. :sweat:

**Comment by earlye** [comment]
Don't underestimate us. Some of us are willing to help, but without a little direction it's really difficult for us to even get started.

**Comment by nightcoding** [comment]
+1 currently a blocker for us

**Comment by Unknwon** [comment]
Whoever watches on this thread, please help test `develop` branch or https://try.gogs.io

**Comment by guillemcanal** [comment]
@Unknwon I'll for sure ;)

**Comment by Techwolf12** [comment]
@Unknwon I will as well!

**Comment by bastiao** [comment]
This is really great news! It seems to work! I will test with more detail for sure.

**Comment by Techwolf12** [comment]
@Unknwon 

* When on master and pressing new pull request in the pull request list or pressing the pull request button on the master branch page, url (/techwolf12/test/compare/master...master) gives a 404, it isn't supposed to make a master to master pull request anyway
* When on dev branch, pressing the pull request icon goes to (/techwolf12/test/compare/master...dev) and gives a 404

**Comment by Unknwon** [comment]
@Techwolf12 link on https://try.gogs.io?

**Comment by Techwolf12** [comment]
@Unknwon The rest on try.gogs.io is usable for me. Can you create a pull request between branches on my repo? https://try.gogs.io/techwolf12/test I've given you collaborator rights

**Comment by Unknwon** [comment]
@Techwolf12 Yes... https://try.gogs.io/techwolf12/test/compare/master...dev

![image](https://cloud.githubusercontent.com/assets/2946214/13540067/5622d052-e222-11e5-9868-5e7b94c87b0a.png)

**Comment by Unknwon** [comment]
@Techwolf12 I might know the problem, let me check.

**Comment by Techwolf12** [comment]
@Unknwon Sure. Please let me know once you want me to test again

**Comment by Techwolf12** [comment]
@Unknwon @guillemcanal 
After being added to https://try.gogs.io/gogogo/testing-repo (Thanks, Guillem!) I can make a pull request. However I do get a 500 upon creation. I suspect it has something to do with the slash in the branch name: https://try.gogs.io/gogogo/testing-repo/compare/master...feat/add-content-to-readme

**Comment by Unknwon** [comment]
@Techwolf12 Updated demo site, you should be able to create PR on your own repository now.

**Comment by Unknwon** [comment]
Also, https://try.gogs.io/gogogo/testing-repo/pulls/1 @Techwolf12

**Comment by Techwolf12** [comment]
@Unknwon Success! https://try.gogs.io/techwolf12/test/pulls/1#issuecomment-620  
However the 500 error is for every branch I try to merge on https://try.gogs.io/gogogo/testing-repo. Can you check in the server logs for me?

**Comment by Unknwon** [comment]
@Techwolf12 error:

```
2016/03/04 16:17:25 [...routers/repo/pull.go:680 CompareAndPullRequestPost()] [E] NewPullRequest: newIssue: Find all labels: Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '))' at line 1
```

**Comment by Techwolf12** [comment]
@Unknwon I am going to try to look into that. An other issue: https://try.gogs.io/techwolf12/test/pulls/2 
Merging it manually from command line because of a conflict, it doesn't resolve the pull request and the changed files/commits are on N/A (and non-clickable)

**Comment by guillemcanal** [comment]
It look good so far, great job! Merge conflict are properly detected.

**Comment by Unknwon** [comment]
@Techwolf12 I think this is expected, you can only view until you solve the conflict and push again.

**Comment by Techwolf12** [comment]
@Unknwon I agree with the part you can only view and needing to merge from command line. However once you do this merge it doesn't show the commits left and you need to manually close it on the PR page. It does however work partially as expected and is not a very blocking issue now.

**Comment by guillemcanal** [comment]
@Unknwon @Techwolf12 I may be mistaken, but I solved a merge conflict on that branch an still, I'm not allowed to merge: https://try.gogs.io/gogogo/testing-repo/pulls/3

**Comment by Unknwon** [comment]
@guillemcanal close and reopen solve the problem, I'll dig into why.

**Comment by guillemcanal** [comment]
@Unknwon Thanks ;)

**Comment by Unknwon** [comment]
@guillemcanal did you use `push --force` to solve the conflict?

**Comment by guillemcanal** [comment]
@Unknwon yes indeed

**Comment by Unknwon** [comment]
@guillemcanal solved.

**Comment by guillemcanal** [comment]
@Unknwon just so you know: I had to logout/login to see the merge pull request button. Maybe it's a side effect caused by the client been updated in the process ;)

**Comment by Unknwon** [comment]
> I had to logout/login to see the merge pull request button

What is this exactly?

**Comment by guillemcanal** [comment]
@Unknwon Well, on the pull request page, before I logged out, the "merge pull request" (I don't really know if it's the exact phrase for that button) where not present. So, I decided to log out and log back in. Then on the "merge pull request" came back.

**Comment by Unknwon** [comment]
> @Unknwon Well, on the pull request page, before I logged out, the "merge pull request" (I don't really know if it's the exact phrase for that button) where not present. So, I decided to log out and log back in. Then on the "merge pull request" came back.

I think refresh the page will do the same job, become it was checking conflict at the moment you send the new pull request. But checking is fast enough to show you result in the next refresh.

**Comment by guillemcanal** [comment]
@Unknwon So I tested another merge conflict pull request, once resolved, It went as expected. Nice work you did there :+1:

**Comment by guillemcanal** [comment]
@Unknwon One last thing: It seems that collaborators have the right to push commits on the master branch, have you any plans to release a feature to prevent such a thing ? Anyway, nice work so far, Go is a really nice language. I'll give it a try ASAP.

**Comment by Unknwon** [comment]
@guillemcanal you can follow https://github.com/gogits/gogs/issues/776

**Comment by guillemcanal** [comment]
@Unknwon thanks, subscribed :)

**Comment by joshfng** [comment]
@Unknwon great work I've been testing locally. One thing I noticed is the activity feed shows "Issue closed" when closing a PR, but it says "Pull Request opened" when opening one.

<img width="745" alt="screen shot 2016-03-04 at 9 00 01 pm" src="https://cloud.githubusercontent.com/assets/302764/13544946/38953370-e24c-11e5-8682-120ee54522d4.png">

**Comment by Unknwon** [comment]
@joshfng good catch!

**Comment by Unknwon** [comment]
@joshfng old ones can't fix, but new ones will look like:

![image](https://cloud.githubusercontent.com/assets/2946214/13549401/19fdaab0-e2d2-11e5-9667-778b361a601c.png)

Now on `develop` branch and demo site.

**Comment by joshfng** [comment]
:+1:

**Comment by moltam** [comment]
On Github you can choose repos at compare. For example a compare on a forked repo by default looks like this:

![fork-compare-default](https://cloud.githubusercontent.com/assets/4578365/13553151/44a66d9e-e37f-11e5-86c7-cf101da6af6c.png)

You can switch to local compare if you click on 'compare across forks' (Same thing happens if you choose your forked repo as base.):

![fork-compare-local](https://cloud.githubusercontent.com/assets/4578365/13553157/85708fee-e37f-11e5-8efb-58fc20c50a63.png)

On the original repo (which was forked) the compare defaults to local compare (But you can also switch to compare between your forks.):

![owner-compare-default](https://cloud.githubusercontent.com/assets/4578365/13553170/088ab4cc-e380-11e5-8225-389e50fddf37.png)

@Unknwon Do you think this could be implemented in Gogs? Not necessary in this issue, but maybe later.

Now in Gogs on a fork you can compare only to the origin repo. I get a 404 error on a fork when I manually change the URL to compare between local branches.

**Comment by Unknwon** [comment]
@moltam it could, but not in near future. :sweat: and out of scope of this thread, you can open another issue to keep the record.

**Comment by Unknwon** [comment]
Close as implemented.

**Comment by Techwolf12** [comment]
Whooo! Nice work @Unknwon :)

**Comment by drsect0r** [comment]
Whooo! Awesome @Unknwon  :+1:

**Comment by sardaukar** [comment]
Great job!! Kudos @Unknwon

**Comment by daniviga** [comment]
Thank you @Unknwon !!!

**Comment by BurakDev** [comment]
Finally, thanks @Unknwon :+1:

**Comment by lucacervasio** [comment]
thanks

**Comment by andreasgruenh** [comment]
I'm using gogs on Version: 0.9.13.0319
Whenever I press the green PR button, i get redirected to /compare/master...master and it shows a 404 error :(
Would really like to use this feature!

**Comment by drsect0r** [comment]
@andreasgruenh Try updating to the latest Gogs version (v0.9.71 at time of writing)

**Comment by andreasgruenh** [comment]
Is it save to upgrade without losing settings or repos?

**Comment by drsect0r** [comment]
@andreasgruenh It should be safe. You can always make a backup of you bare repositories and database. (you should git bundle when backing up repositories)

**Comment by bkcsoft** [comment]
you can use `./gogs dump` as well to make a backup of everything

**Comment by andreasgruenh** [comment]
I just updated to Gogs Version: 0.9.71.0809 and I still get the 404 error when clicking on one of the pull-request buttons. Any chance my reverse proxy causes this?

**Comment by dpheenatigala** [comment]
I have had problems with trying to create pull requests with gogs version 0.9.97.0901 where branches have hyphens in them. The below log is created when i try to compare "resource-allocation-tasks" and "develop" branch.

```
2016/10/28 11:32:22 [...uters/repo/commit.go:230 CompareDiff()] [E] GetCommit: exit status 128 - fatal: ambiguous argument 'resource': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

P.S. The compare url pointed to by the "octicon-git-compare" doesn't change when the branch that is viewed is changed. This means the link always goes to "[user]/[repo]/compare/develop...develop" despite which branch i am browsing currently.

**Comment by M2shad0w** [comment]
Whooo! Awesome @Unknwon 👍
